### PR TITLE
GH-36744: [Python][Packaging] Add upper pin for cython<3 to pyarrow build dependencies

### PR DESCRIPTION
**Pull request by jorisvandenbossche**
### Rationale for this change

Although we already fixed some cython 3 build issues (https://github.com/apache/arrow/pull/34726), some new have been introduced, which we are seeing now cython 3 is released (https://github.com/apache/arrow/issues/36730)

Adding an upper pin (<3) for the release, so we have more time (the full 14.0 release cycle) to iron out issues.
* Closes: #36744

### CODE REVIEW

**Comment by pitrou** [comment]
@github-actions crossbow submit -g python

**Review by raulcd — COMMENTED** [review]
aren't the changes on setup.py required?
```
diff --git a/python/setup.py b/python/setup.py
index f06cb5a..dc52967 100755
--- a/python/setup.py
+++ b/python/setup.py
@@ -40,8 +40,9 @@ import Cython
 # Check if we're running 64-bit Python
 is_64_bit = sys.maxsize > 2**32
 
-if Cython.__version__ < '0.29.31':
-    raise Exception('Please upgrade to Cython 0.29.31 or newer')
+if Cython.__version__ < '0.29.31' or Cython.__version__ >= '3.0':
+    raise Exception(
+        'Please update your Cython version. Supported Cython >= 0.29.31, < 3.0')
 
 setup_dir = os.path.abspath(os.path.dirname(__file__))
 
@@ -491,7 +492,7 @@ setup(
                                  'pyarrow/_generated_version.py'),
         'version_scheme': guess_next_dev_version
     },
-    setup_requires=['setuptools_scm', 'cython >= 0.29.31'] + setup_requires,
+    setup_requires=['setuptools_scm', 'cython >= 0.29.31,<3'] + setup_requires,
```

**Comment by raulcd** [comment]
@github-actions crossbow submit -g wheel

**Comment by raulcd** [comment]
The failure on `example-python-minimal-build-fedora-conda` seems related to:
https://github.com/apache/arrow/blob/main/ci/conda_env_python.txt#L21
We should pin Cython there too

**Comment by jorisvandenbossche** [comment]
@github-actions crossbow submit example-python-minimal-build-fedora-conda example-python-minimal-build-ubuntu-venv

**Comment by jorisvandenbossche** [comment]
> aren't the changes on setup.py required?

No idea what this `setup_requires` is actually still doing (now there is pyproject.toml), we should probably check to remove that.

It's not really "required", but indeed nicer to update the error message (it would also have been clearer on the conda build). On the other hand, I am not sure it is necessary to fully disallow building with cython 3, for example for local development. But assuming we remove the pin soon again on the main branch, I added it for now.

> [about conda env] We should pin Cython there too

We actually shouldn't, if we would update our install commands in CI to use pip instead of calling setup.py directly. But that's for another issue ;)

**Comment by raulcd** [comment]
to fix the `Source Release and Merge Script` failures:
```
diff --git a/.github/workflows/dev.yml b/.github/workflows/dev.yml
index 7c2437f..bfcefcb 100644
--- a/.github/workflows/dev.yml
+++ b/.github/workflows/dev.yml
@@ -103,7 +103,7 @@ jobs:
         shell: bash
         run: |
           gem install test-unit
-          pip install cython setuptools six pytest jira
+          pip install cython<3 setuptools six pytest jira
       - name: Run Release Test
```

**Comment by jorisvandenbossche** [comment]
Thanks! Pushed (and that's another place where we should try to avoid installing cython manually, I think)

**Review by pitrou — APPROVED** [review]

**Comment by pitrou** [comment]
@github-actions crossbow submit -g python

**Comment by pitrou** [comment]
@jorisvandenbossche @raulcd Can we merge this once CI passes?

**Review by raulcd — APPROVED** [review]
I'll cherry pick for 13.0.0 once merged

**Pull request merged**
